apisix/discovery/nacos.lua

```diff
@@ -0,0 +1,326 @@
+    else
+        token_param = ""
+    end
+end
+
+
+local function service_info()
+    local host = local_conf.discovery and local_conf.discovery.nacos
+                 and local_conf.discovery.nacos.host
+    if not host then
+        log.error("do not set nacos.host")
+        return
+    end
+
+    local username, password
+    -- TODO Add health check to get healthy nodes.
+    local url = host[math_random(#host)]
+    local auth_idx = str_find(url, "@")
+    if auth_idx then
+        local protocol_idx = str_find(url, "://")
+        local protocol = string_sub(url, 1, protocol_idx + 2)
+        local user_and_password = string_sub(url, protocol_idx + 3, auth_idx - 1)
+        local arr = ngx_re.split(user_and_password, ":")
+        if #arr == 2 then
+            username = arr[1]
+            password = arr[2]
+        end
+        local other = string_sub(url, auth_idx + 1)
+        url = protocol .. other
+    end
+
+    if local_conf.discovery.nacos.prefix then
+        url = url .. local_conf.discovery.nacos.prefix
+    end
+
+    if str_byte(url, #url) ~= str_byte("/") then
+        url = url .. "/"
+    end
+
+    base_uri = url
+    refresh_token_param(username, password)
+end
+
+
+local function get_page_service(infos, base_uri, token_param, page_num)
+    local path = str_format(service_list_path, page_num, page_size) .. token_param
+    local data, err = get_url(base_uri, path)
+    if err then
+        return data, err, path
+    end
+
+    for _, service_name in ipairs(data.doms) do
+        core.table.insert(infos, service_name)
+    end
+    return data, err, path
+end
+
+
+local function iter_and_add_service_info(infos, base_uri, token_param)
+    local data, err, path = get_page_service(infos, base_uri, token_param, 1)
+    if err then
+        log.error("get_url:" .. path .. " err:" .. err)
+        return
+    end
+
+    local maxPage = math.ceil(data.count / page_size)
+    if maxPage == 0 then
+        return
+    end
+
+    if maxPage > 1 then
+        for i = 2, maxPage do
+            get_page_service(infos, base_uri, token_param, i)
+        end
+    end
+end
+
+
+local function get_services(base_uri, token_param)
+    local infos = core.table.new(0, 0)
+    iter_and_add_service_info(infos, base_uri, token_param)
+    return infos
+end
+
+
+local function fetch_full_registry(premature)
+    if premature then
+        return
+    end
+
+    local up_apps = core.table.new(0, 0)
+    service_info()
+    if not base_uri then
+        applications = up_apps
+        return
+    end
+
+    local infos = get_services(base_uri, token_param)
+    if #infos == 0 then
+        applications = up_apps
+        return
+    end
+
+    local data, err
+    for _, service_name in ipairs(infos) do
+        data, err = get_url(base_uri, instance_list_path .. service_name .. token_param)
+        if err then
+            log.error("get_url:" .. instance_list_path .. " err:" .. err)
+            applications = up_apps
```

Comment:
IMHO, when something bad happened, we should keep the `application` unchanged. Stale configuration is better than no configuration.